### PR TITLE
config/containers-and-vms/lxc: document incus

### PR DESCRIPTION
**Pull request by duskmoss**
Incus is the current community tool for easy LXC administration. LXD has been moved to canonical, relicensed (Apache 2.0 to AGPL3) and contributors are asked to sign a CLA. 

LXD package in repos is out of date, but incus is up to date. I thought we should have docs for this package. 

For now I've left the LXD section in as ahesford recommended on IRC.

### CODE REVIEW

File: src/config/containers-and-vms/lxc.md [hunk 1]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
 create each container. Trying to fix permissions on a container created with the
 wrong map is possible, but inconvenient.
 
+## Incus
+
+Incus provides an alternative interface to LXC's `lxc-*` utilities. However, it
+does not require the configuration described in [the previous section](#lxc).
+
+Edit `/etc/rc.conf` and set the `CGROUP_MODE` variable to `unified`. Install the
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:127]:
```suggestion
In `/etc/rc.conf`, set the `CGROUP_MODE` variable to `unified`. Install the
```

File: src/config/containers-and-vms/lxc.md [hunk 2]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+`incus` package, and [enable](../services/index.md#enabling-services) the
+`incus` service.
+
+Incus users must be belong to the `_incus-admin` group.
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:131]: what are "incus users"? users that can perform admin actions in incus?

File: src/config/containers-and-vms/lxc.md [hunk 3]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+
+Optionally enable some users to have limited acces to incus as described
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:133]:
```suggestion
Optionally, some users can be given limited access to Incus as described
```

File: src/config/containers-and-vms/lxc.md [hunk 4]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+[here](https://linuxcontainers.org/incus/docs/main/howto/projects_confine/#confine-projects-to-specific-incus-users).
+Add users to the `_incus` group, and enable the `incus-user` service
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:135]:
```suggestion
Add users to the `_incus` group, and enable the `incus-user` service.
```

File: src/config/containers-and-vms/lxc.md [hunk 5]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+Note that `incus-user` will initialize the default incus profile when it is
+started. To avoid default configuration initialize incus for youself with `incus
+admin init` before enabling `incus-user`.
+
+**WARNING:** `incus-user` will also replace the networking config of the default
+profile if it appears invalid. runit does not have socket activation so this
+happens when when the machine boots instead of when a user attempts to access
+it. If the default profile uses a [bridge
+interface](../network/index.md#bridge-interfaces) `incus-user` may replace it on
+boot.
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:146]:
```suggestion
> Warning: `incus-user` will also replace the networking config of the default
> profile if it appears invalid. runit does not have socket activation so this
> happens when when the machine boots instead of when a user attempts to access
> it. If the default profile uses a [bridge
> interface](../network/index.md#bridge-interfaces) `incus-user` may replace it on
> boot.
```

File: src/config/containers-and-vms/lxc.md [hunk 6]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+
+To migrate existing LXD setups to incus use the 'lxd-to-incus' binary from the
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:148]:
```suggestion
To migrate existing LXD setups to Incus, use the `lxd-to-incus` tool from the
```

File: src/config/containers-and-vms/lxc.md [hunk 7]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+`incus-tools` package as described
+[here](https://linuxcontainers.org/incus/docs/main/howto/server_migrate_lxd/).
+
+To migrate existing LXC containers to incus use the 'lxc-to-incus' binary from
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:152]:
```suggestion
To migrate existing LXC containers to Incus use the `lxc-to-incus` script from
```

**Review by classabbyamp — COMMENTED** [review]
mostly copyediting review as I don't know much about incus

File: src/config/containers-and-vms/lxc.md (hunk 2 above)
Comment (0x5c) [reply src/config/containers-and-vms/lxc.md:131]: Upstream docs do say that the `_incus-admin` group should only be given to users who should be able to do system-wide incus operations without root. The normal recommended way to use incus as a normal user (other than setup of the daemon) is with the plain `_incus` group.

File: src/config/containers-and-vms/lxc.md [hunk 8]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+`incus` service.
+
+Incus users must be belong to the `_incus-admin` group.
+
+Optionally enable some users to have limited acces to incus as described
+[here](https://linuxcontainers.org/incus/docs/main/howto/projects_confine/#confine-projects-to-specific-incus-users).
+Add users to the `_incus` group, and enable the `incus-user` service
+
+Note that `incus-user` will initialize the default incus profile when it is
+started. To avoid default configuration initialize incus for youself with `incus
```
Comment (0x5c) [inline src/config/containers-and-vms/lxc.md:138]:
```suggestion
started. To avoid default configuration initialize incus for yourself with `incus
```

**Review by 0x5c — COMMENTED** [review]
> Note that `incus-user` will initialize the default incus profile when it is started. To avoid default configuration initialize incus for youself with `incus admin init` before enabling `incus-user`.

I spent multiple days trying to figure out what was going on with incus automatically creating a new zfs pool, thanks for saving me from going mad :D

File: src/config/containers-and-vms/lxc.md (hunk 2 above)
Comment (duskmoss) [reply src/config/containers-and-vms/lxc.md:131]: @classabbyamp yes, that's what I meant by incus users. all actions are admin action in normal operation as it's roughly a hypervisor. Open to rephrasing.

@0x5c I don't think that's a fair characterization The normal way is to have access to everything in the incus admin. adding to just the incus group gets a user access to just a personalized project space as described here, as an optional isolation strategy. 

https://linuxcontainers.org/incus/docs/main/explanation/projects/#projects-confined

Having an admin is essential, having limited access is optional, is my view I wrote into the docs. 


or riffed on briefly here 

https://linuxcontainers.org/incus/docs/main/tutorial/first_steps/

    "incus allows basic user access, no configuration and all actions restricted to a per-user project."

   " incus-admin allows full control over Incus."

Comment (0x5c) [reply src/config/containers-and-vms/lxc.md:131]: The [upstream documentation page for install instructions](https://linuxcontainers.org/incus/docs/main/installing/) has this, which points out that only root access is actually needed to perform system-wide operations
> The root user and all members of the incus-admin group can interact with the local daemon.

It also includes
> :fire: Important
> Local access to Incus through the Unix socket always grants full access to Incus. This includes the ability to attach file system paths or devices to any instance as well as tweak the security features on any instance.
> Therefore, you should only give such access to users who you’d trust with root access to your system.

Additionally, the [About Security](https://linuxcontainers.org/incus/docs/main/explanation/security/) page states
> - Restrict access to the Incus daemon and the remote API.
> - Do not use privileged containers unless required. If you use privileged containers, put appropriate security measures in place. See the [LXC security page](https://linuxcontainers.org/lxc/security/) for more information.

The above mentioned LXC Security page has this to say about privileged and unprivileged containers
> The former can be thought as old-style containers, they're not safe at all and should only be used in environments where unprivileged containers aren't available and where you would trust your container's user with root access to the host.

Critically, `_incus-admin` is equivalent to unrestricted passwordless sudo access.

While I agree that the `_incus-admin` group should be mentioned, saying that users "must" be in that group does not sound like a reasonable *default* recommendation, especially when prefixing admin commands with `sudo`/`doas` would do the same without the problem of granting passwordless root access.
In other words, while the Void documentation should not be there to criticise security choices a user could deem safe for their setup, the documentation shouldn't be making the assumption of what's "super safe enough" in what it recommends, especially when upstream is quite clear about the risks.

To me the proper way to handle this is to indicate that root access in needed to perform incus administrative operations, with a mention of `_incus-admin` as an alternative along with its caveats. Further details may possibly be more appropriate for the upstream documentation, which can be linked to.

Comment (classabbyamp) [reply src/config/containers-and-vms/lxc.md:131]: that makes sense to me

Comment (0x5c) [reply src/config/containers-and-vms/lxc.md:131]: Sound good to me

**Comment by 0x5c** [comment]
There's two remaining bits of documentation in the `README.voidlinux` file that if added here would render it completely unnecessary.

> The root user should also be assigned a range of subordinate user and group IDs to be mapped in
containers.
> For example, the command
> ```
> usermod --add-subuids 1000000-1065535 \
> 	--add-subgids 1000000-1065535 root
> ```
> will assign a range of 65,536 IDs for users and groups.

The sub{u,g}id stuff is mentioned vaguely (with different values and the manual method) in the LXC section, but there's no indication that one would need to follow the instruction in the LXC section, nor good indication at to what it means in practice Maybe this could be refactored into a section that both the LXC and Incus sections link to.


> Optional dependencies:
> 
> To run virtual machines, install `qemu` and `edk2-ovmf`.
> To run OCI containers, install `spokeo` and `umoci`.

**Comment by duskmoss** [comment]
> There's two remaining bits of documentation in the `README.voidlinux` file that if added here would render it completely unnecessary.

I didn't include these because it's actually not neccesary, these are setup in void by default. I did make a pr to remove from the readme but maintainer wanted to keep.

**Comment by 0x5c** [comment]
Interesting, the only range that was setup without my intervention was for my user, not root. What part of void would be setting that up automatically?

**Comment by duskmoss** [comment]
https://github.com/void-linux/void-packages/blob/master/srcpkgs/base-files/files/subgid

**Comment by 0x5c** [comment]
Weirdly, that was added 5 years ago and my install is not that old, I guess something to investigate separately from this issue

**Comment by duskmoss** [comment]
I'll try and confirm at some point, but eitherway, it's in the readme. 

I think optional dependencies should stay in the readme, it's much easier to update them when the package is updated than keeping the documentation in sync. And it's the standard place for void, like with the steam package. 

I'll get the edits talked about here worked on either today or tomorrow.

File: src/config/containers-and-vms/lxc.md [hunk 9]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+## Incus
+
+Incus provides an alternative interface to LXC's `lxc-*` utilities. However, it
+does not require the configuration described in [the previous section](#lxc).
+
+In `/etc/rc.conf`, set the `CGROUP_MODE` variable to `unified`. Install the
+`incus` package, and [enable](../services/index.md#enabling-services) the
+`incus` service.
+
+Add users who should have full control over Incus to the '_incus-admin' group.
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:131]:
```suggestion
Add users who should have full control over Incus to the `_incus-admin` group.
```

File: src/config/containers-and-vms/lxc.md [hunk 10]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+
+Optionally, some users can be given limited access to Incus as described
+[here](https://linuxcontainers.org/incus/docs/main/howto/projects_confine/#confine-projects-to-specific-incus-users).
+Add users to the `_incus` group, and enable the `incus-user` service.
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:135]:
```suggestion
Add these users to the `_incus` group, and enable the `incus-user` service.
```

File: src/config/containers-and-vms/lxc.md [hunk 11]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+
+Note that `incus-user` will initialize the default Incus profile when it is
+started. To avoid default configuration initialize Incus for yourself with
+`incus admin init` before enabling `incus-user`.
+
+> Warning: `incus-user` will also replace the networking config of the default
+> profile if it appears invalid. runit does not have socket activation so this
+> happens when when the machine boots instead of when a user calls `incus`. If
+> the default profile uses a [bridge
+> interface](../network/index.md#bridge-interfaces) `incus-user` may replace it
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:145]:
```suggestion
> interface](../network/index.md#bridge-interfaces) that `incus-user` may replace
```

File: src/config/containers-and-vms/lxc.md [hunk 12]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
+> on boot.
+
+To migrate existing LXD setups to Incus, use the `lxd-to-incus` tool from the
+`incus-tools` package as described
+[here](https://linuxcontainers.org/incus/docs/main/howto/server_migrate_lxd/).
+
+To migrate existing LXC containers to Incus use the `lxc-to-incus` script from
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:152]:
```suggestion
To migrate existing LXC containers to Incus, use the `lxc-to-incus` script from
```

File: src/config/containers-and-vms/lxc.md [hunk 13]
```diff
@@ -114,6 +119,40 @@ access to other containers. To isolate containers from each other, alter the
...
 wrong map is possible, but inconvenient.
 
+## Incus
+
+Incus provides an alternative interface to LXC's `lxc-*` utilities. However, it
+does not require the configuration described in [the previous section](#lxc).
+
+In `/etc/rc.conf`, set the `CGROUP_MODE` variable to `unified`. Install the
+`incus` package, and [enable](../services/index.md#enabling-services) the
+`incus` service.
```
Comment (classabbyamp) [inline src/config/containers-and-vms/lxc.md:129]: maybe we could mention looking in README.voidlinux for optional dependencies? like `Some parts of Incus require optional dependencies, see [README.voidlinux](../package-documentation/index.md).` (link might not be 100% right)

Comment (duskmoss) [reply src/config/containers-and-vms/lxc.md:129]: Sounds good

File: src/config/containers-and-vms/lxc.md (hunk 11 above)
Comment (duskmoss) [reply src/config/containers-and-vms/lxc.md:145]: I don't think this reads correctly.  "If there is a a bridge that incus may replace on boot" isn't a complete sentence.

Comment (classabbyamp) [reply src/config/containers-and-vms/lxc.md:148]: oops, got messed up by link/line breaks. maybe a comma after the link instead:
```suggestion
> interface](../network/index.md#bridge-interfaces), `incus-user` may replace it
```

**Comment by classabbyamp** [comment]
> I think optional dependencies should stay in the readme, it's much easier to update them when the package is updated than keeping the documentation in sync.

a pr to void-docs updating that kind of thing can be merged quickly, but it is also valuable to have them in README.voidlinux for history.

Can we put them in both places?

**Comment by duskmoss** [comment]
Sure

**Comment by classabbyamp** [comment]
and btw, commit message should be `config/containers-and-vms/lxc: add incus`

**Comment by duskmoss** [comment]
> and btw, commit message should be `config/containers-and-vms/lxc: document incus` or something like that

Thanks! Maybe the contribution docs should be more clear about what you want for <filename> because path, and file name with  extension removed is not something I would have gotten from <filename>
